mws/sitesmanagement/cronjobs.py

```diff
@@ -309,3 +310,67 @@ def reject_or_accepted_old_domain_names_requests():
                                    "%s days.") % (grace_days,))
         else:
             domain_name.accept_it()
+
+@shared_task(base=ScheduledTaskWithFailure)
+def validate_domains():
+    '''
+    Iterate over DomainName objects and set them to:
+     - global if they are visible to (currently) Google's nameservers
+     - private if they are onl`y available to the Cambridge nameservers
+     - deleted if they are visible to none of the above.
+    '''
+    active_states = ['accepted', 'private', 'global', 'deleted']
+
+    for domainname in DomainName.objects.filter(status__in=active_states):
+            status = domainname.validate()
+            if status != domainname.status:
+                domainname.status = status
+                domainname.save()
+
+@shared_task(base=ScheduledTaskWithFailure)
+def expire_domains():
+    '''
+    Periodically send messages to domain administrators about deleted domains and
+    delete them after grace_days
+    '''
+    grace_days = settings.MWS_DOMAIN_NAME_GRACE_DAYS
+    notify_days = list(range(1, grace_days, grace_days//3+1))
+    support_email = settings.EMAIL_MWS3_SUPPORT
+    for domainname in DomainName.objects.filter(status='deleted').annotate(expired=(now() - F('updated_at'))):
```

Comment:
'external' and 'special' will never get to the status 'deleted' as we are not calling validate() on them. Line 325 only calls validate for domains in ['accepted', 'private', 'global', 'deleted'] states.

Reply:
Also we are calling validate on line 325 for domain names with status 'deleted'. That means that the update_at will get updated every time we call it, which will cause the expired annotation not to represent reality.

Reply:
added 'external' and 'special' to active_states list. validate only saves a model if its status changes, so the annotation should work correctly.

Reply:
oh yes! I missed that